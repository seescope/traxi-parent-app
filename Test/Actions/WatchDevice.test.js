jest.mock('../../App/Actions/SaveProfile', () =>
  jest.fn(() => Promise.resolve()));
global.Promise = require.requireActual('promise');

import { mockOn } from 'firebase-old';
import { TEST_KID } from '../Mocks';
import Analytics from 'react-native-analytics';
import watchDevice from '../../App/Actions/WatchDevice';
import saveProfile from '../../App/Actions/SaveProfile';

it('registers a listener with Firebase that dispatches a DEVICE_UPDATED action', () => {
  const mockGetState = () => ({ selectedKid: TEST_KID });
  const mockDispatch = jest.fn();

  // Exercise
  watchDevice()(mockDispatch, mockGetState);

  // Assert
  expect(mockOn.mock.calls).toMatchSnapshot();
  expect(mockDispatch.mock.calls).toMatchSnapshot();
});

it('handles when the device has been installed', () => {
  const mockGetState = () => ({
    selectedKid: TEST_KID,
    profile: { kids: ['First Kid'] },
  });
  const mockDispatch = jest.fn();
  mockOn.setData({ ...TEST_KID, status: 'INSTALLED' });

  return watchDevice()(mockDispatch, mockGetState).then(() => {
    expect(mockOn.mock.calls).toMatchSnapshot();
    expect(mockDispatch.mock.calls).toMatchSnapshot();
    expect(Analytics.track.mock.calls).toMatchSnapshot();
  });
});

it('handles duplicate kids', () => {
  const TEST_KIDS = [{ UUID: 'Kid 1' }, { UUID: 'Kid 2' }, { UUID: 'Kid 2' }];

  const mockGetState = () => ({
    selectedKid: TEST_KID,
    profile: { kids: TEST_KIDS },
  });

  const mockDispatch = jest.fn();
  mockOn.setData({ ...TEST_KID, status: 'INSTALLED' });

  return watchDevice()(mockDispatch, mockGetState).then(() => {
    expect(mockOn.mock.calls).toMatchSnapshot();
    expect(mockDispatch.mock.calls).toMatchSnapshot();
    expect(saveProfile.mock.calls).toMatchSnapshot();
  });
});
