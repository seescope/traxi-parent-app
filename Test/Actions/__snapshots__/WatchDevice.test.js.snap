exports[`test handles duplicate kids 1`] = `
Array [
  Array [
    "value",
    [Function],
    [Function],
  ],
  Array [
    "value",
    [Function],
    [Function],
  ],
  Array [
    "value",
    [Function],
    [Function],
  ],
]
`;

exports[`test handles duplicate kids 2`] = `
Array [
  Array [
    Object {
      "selectedKid": Object {
        "UUID": "a5f1bc36-89f5-4212-98c6-e92797a2a220",
        "avatarURL": "https://s-media-cache-ak0.pinimg.com/736x/24/48/12/2448127e36c0e72de41aa768322639a3.jpg",
        "name": "Edith Crawley",
        "status": "INSTALLED",
      },
      "type": "DEVICE_UPDATED",
    },
  ],
  Array [
    Object {
      "profile": Object {
        "kids": Array [
          Object {
            "UUID": "Kid 1",
          },
          Object {
            "UUID": "Kid 2",
          },
          Object {
            "UUID": "a5f1bc36-89f5-4212-98c6-e92797a2a220",
            "avatarURL": "https://s-media-cache-ak0.pinimg.com/736x/24/48/12/2448127e36c0e72de41aa768322639a3.jpg",
            "name": "Edith Crawley",
            "status": "INSTALLED",
          },
        ],
      },
      "type": "LOGGED_IN",
    },
  ],
]
`;

exports[`test handles duplicate kids 3`] = `
Array [
  Array [
    Object {
      "kids": Array [
        "First Kid",
        Object {
          "UUID": "a5f1bc36-89f5-4212-98c6-e92797a2a220",
          "avatarURL": "https://s-media-cache-ak0.pinimg.com/736x/24/48/12/2448127e36c0e72de41aa768322639a3.jpg",
          "name": "Edith Crawley",
          "status": "INSTALLED",
        },
      ],
    },
  ],
  Array [
    Object {
      "kids": Array [
        Object {
          "UUID": "Kid 1",
        },
        Object {
          "UUID": "Kid 2",
        },
        Object {
          "UUID": "a5f1bc36-89f5-4212-98c6-e92797a2a220",
          "avatarURL": "https://s-media-cache-ak0.pinimg.com/736x/24/48/12/2448127e36c0e72de41aa768322639a3.jpg",
          "name": "Edith Crawley",
          "status": "INSTALLED",
        },
      ],
    },
  ],
]
`;

exports[`test handles when the device has been installed 1`] = `
Array [
  Array [
    "value",
    [Function],
    [Function],
  ],
  Array [
    "value",
    [Function],
    [Function],
  ],
]
`;

exports[`test handles when the device has been installed 2`] = `
Array [
  Array [
    Object {
      "selectedKid": Object {
        "UUID": "a5f1bc36-89f5-4212-98c6-e92797a2a220",
        "avatarURL": "https://s-media-cache-ak0.pinimg.com/736x/24/48/12/2448127e36c0e72de41aa768322639a3.jpg",
        "name": "Edith Crawley",
        "status": "INSTALLED",
      },
      "type": "DEVICE_UPDATED",
    },
  ],
  Array [
    Object {
      "profile": Object {
        "kids": Array [
          "First Kid",
          Object {
            "UUID": "a5f1bc36-89f5-4212-98c6-e92797a2a220",
            "avatarURL": "https://s-media-cache-ak0.pinimg.com/736x/24/48/12/2448127e36c0e72de41aa768322639a3.jpg",
            "name": "Edith Crawley",
            "status": "INSTALLED",
          },
        ],
      },
      "type": "LOGGED_IN",
    },
  ],
]
`;

exports[`test registers a listener with Firebase that dispatches a DEVICE_UPDATED action 1`] = `
Array [
  Array [
    "value",
    [Function],
    [Function],
  ],
]
`;

exports[`test registers a listener with Firebase that dispatches a DEVICE_UPDATED action 2`] = `
Array [
  Array [
    Object {
      "selectedKid": Object {
        "UUID": "a5f1bc36-89f5-4212-98c6-e92797a2a220",
        "avatarURL": "https://s-media-cache-ak0.pinimg.com/736x/24/48/12/2448127e36c0e72de41aa768322639a3.jpg",
        "name": "Edith Crawley",
      },
      "type": "DEVICE_UPDATED",
    },
  ],
]
`;
